include/clang/AST/Expr.h

```diff
@@ -4514,6 +4514,120 @@ class GenericSelectionExpr : public Expr {
   friend class ASTStmtReader;
 };
 
+/// \brief Represents a Checked C bounds expression.
+class BoundsExpr : public Expr {
+private:
+  SourceLocation StartLoc, RParenLoc;
+
+public:
+  BoundsExpr(StmtClass StmtClass, SourceLocation StartLoc, SourceLocation RParenLoc)
+    : Expr(StmtClass, QualType(),  VK_RValue, OK_Ordinary, false,
+           false, false, false), StartLoc(StartLoc), RParenLoc(RParenLoc) {
+  }
+
+  explicit BoundsExpr(StmtClass StmtClass, EmptyShell Empty) :
+    Expr(StmtClass, Empty) {}
+
+  SourceLocation getStartLoc() { return StartLoc; }
+  void setStartLoc(SourceLocation Loc) { StartLoc = Loc; }
+  SourceLocation getRParenLoc() { return RParenLoc; }
+  void setRParenLoc(SourceLocation Loc) { RParenLoc = Loc; }
+
+  SourceLocation getLocStart() const LLVM_READONLY { return StartLoc; }
+  SourceLocation getLocEnd() const LLVM_READONLY { return RParenLoc; }
+};
+
+/// \brief Represents a Checked C nullary bounds expression.
```

Comment:
I might need to add bounds(Any) back for internal IR processing/checking.   I agree that we can specialize this if we don’t end up needing the generality.

From: Reuben Olinsky [mailto:notifications@github.com]
Sent: Tuesday, June 28, 2016 5:07 PM
To: Microsoft/checkedc-clang <checkedc-clang@noreply.github.com>
Cc: David Tarditi <dtarditi@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Microsoft/checkedc-clang] Extend clang IR with bounds expressions and parse bounds expressions for parameters. (#8)


In include/clang/AST/Expr.h<https://github.com/Microsoft/checkedc-clang/pull/8#discussion_r68865825>:

> +           false, false, false), StartLoc(StartLoc), RParenLoc(RParenLoc) {

> +  }

> +

> +  explicit BoundsExpr(StmtClass StmtClass, EmptyShell Empty) :

> +    Expr(StmtClass, Empty) {}

> +

> +  SourceLocation getStartLoc() { return StartLoc; }

> +  void setStartLoc(SourceLocation Loc) { StartLoc = Loc; }

> +  SourceLocation getRParenLoc() { return RParenLoc; }

> +  void setRParenLoc(SourceLocation Loc) { RParenLoc = Loc; }

> +

> +  SourceLocation getLocStart() const LLVM_READONLY { return StartLoc; }

> +  SourceLocation getLocEnd() const LLVM_READONLY { return RParenLoc; }

> +};

> +

> +/// \brief Represents a Checked C nullary bounds expression.

Will we ever use this for something other than 'None' bounds? If not, should we make this less generic and simply call it NoneBoundsExpr?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Microsoft/checkedc-clang/pull/8/files/340bbc6733cc1e11b2f0822b2bbc4f7e32fd87b9#r68865825>, or mute the thread<https://github.com/notifications/unsubscribe/ALSuCgytvg4JkS7HC1bNNmtPIDJ72PWvks5qQbc0gaJpZM4I_o4B>.